pkg/apis/pipeline/v1beta1/pipeline_validation.go

```diff
@@ -291,40 +291,68 @@ func validatePipelineContextVariables(tasks []PipelineTask) *apis.FieldError {
 			if ps, ok := GetVarSubstitutionExpressionsForParam(param); ok {
-				for _, p := range ps {
-					if strings.HasPrefix(p, "tasks.") && strings.HasSuffix(p, ".status") {
-						errs = errs.Also(apis.ErrInvalidValue(fmt.Sprintf("pipeline tasks can not refer to execution status of any other pipeline task"),
-							"value").ViaFieldKey("params", param.Name).ViaFieldIndex("tasks", idx))
+				// validate tasks.pipelineTask.status if this expression is not a result reference
+				if !LooksLikeContainsResultRefs(ps) {
+					for _, p := range ps {
+						// check for a prefix and suffix
+						if strings.HasPrefix(p, "tasks.") && strings.HasSuffix(p, ".status") {
+							errs = errs.Also(apis.ErrInvalidValue(fmt.Sprintf("pipeline tasks can not refer to execution status of any other pipeline task"),
+								"value").ViaFieldKey("params", param.Name).ViaFieldIndex("tasks", idx))
+						}
 					}
 				}
 			}
 		}
+	}
+	return errs
+}
+
+// validate finally tasks accessing execution status of a dag task specified in the pipeline
+// $(tasks.pipelineTask.status) is invalid if pipelineTask is not defined as a dag task
+func validateExecutionStatusVariablesInFinally(tasks []PipelineTask, finally []PipelineTask) (errs *apis.FieldError) {
+	// creating a list of pipelineTask names to validate tasks.<name>.status
+	pipelineRunTasksContextNames := sets.String{}
+	for _, t := range tasks {
 		pipelineRunTasksContextNames.Insert(t.Name)
 	}
 
-	// validate finally tasks accessing execution status of a dag task specified in the pipeline
-	var paramValues []string
-	for _, t := range finallyTasks {
+	for idx, t := range finally {
 		for _, param := range t.Params {
-			paramValues = append(paramValues, param.Value.StringVal)
-			paramValues = append(paramValues, param.Value.ArrayVal...)
-		}
-	}
-	for _, paramValue := range paramValues {
-		if strings.HasPrefix(stripVarSubExpression(paramValue), "tasks.") && strings.HasSuffix(stripVarSubExpression(paramValue), ".status") {
-			errs = errs.Also(substitution.ValidateVariablePS(paramValue, "tasks", "status", pipelineRunTasksContextNames).ViaField("value"))
+			// retrieve a list of substitution expression from a param
+			if ps, ok := GetVarSubstitutionExpressionsForParam(param); ok {
+				// validate tasks.pipelineTask.status if this expression is not a result reference
+				if !LooksLikeContainsResultRefs(ps) {
+					for _, p := range ps {
+						// check for a prefix and suffix
+						if strings.HasPrefix(p, "tasks.") && strings.HasSuffix(p, ".status") {
+							pt := strings.TrimSuffix(strings.TrimPrefix(p, "tasks."), ".status")
+							// report an error if the task name does not exist in the list of dag tasks
+							if !pipelineRunTasksContextNames.Has(pt) {
+								errs = errs.Also(apis.ErrInvalidValue(fmt.Sprintf("pipeline task %s is not defined in the pipeline", pt),
+									"value").ViaFieldKey("params", param.Name).ViaFieldIndex("finally", idx))
```

Comment:
can be in a helper func as it is re-used

Reply:
yup definitely, created a helper function.